File: d5/ruby/d5.rb
class Stepper
  attr_reader :steps, :instruction, :tape, :config
  def initialize(tape, opts = {})
    @steps = 0
    @instruction = 0
    @tape = tape.split(/\n+/).map &:to_i
    @config = {
      strange: false,
      strange_offset: 3
    }.merge opts
  end

  def walk!
    while instruction < tape.length
      move tape[instruction]
    end

    puts "Walked #{steps} #{config[:strange] ? 'strange' : 'regular'} steps."
    puts "Final instruction is #{instruction} on a #{tape.length} tape."
  end

  private
  def move(distance = 0)
    tape[instruction] += strange_move?(distance) ? -1 : 1
    @steps += 1
    @instruction += distance
  end

  def strange_move?(distance)
    config[:strange] && distance >= config[:strange_offset]
  end
end

test_tape = "0\n3\n0\n1\n-3"
Stepper.new(test_tape).walk!
Stepper.new(test_tape, strange: true).walk!

Stepper.new(File.read("input")).walk!
Stepper.new(File.read("input"), strange: true).walk!

File: d5/elixir/lib/d5.ex
defmodule D5 do
  @doc """
  iex> File.read!("input") |> D5.walk
  372671

  iex> D5.walk "0\\n3\\n0\\n1\\n-3"
  5
  """
  def walk(tape), do: walk(prep(tape), {0, 0})
  def walk(tape, {steps, instruction}) when instruction >= tuple_size(tape), do: steps
  def walk(tape, {steps, instruction}) do
    current = elem tape, instruction
    tape = tape
      |> Tuple.delete_at(instruction)
      |> Tuple.insert_at(instruction, current + 1)

    walk(tape, {steps + 1, current + instruction})
  end

  defp prep(tape) do
    tape
      |> String.trim
      |> String.split(~r/\n+/)
      |> Enum.map(&String.to_integer/1)
      |> List.to_tuple
  end
end
